Feature: Data Preparation

martinctc opened the pull request:
# Summary
This branch adds any data preparation scripts for Google Sheets, prior to feeding into Shiny app. Runs once (not run during deployment).

# Changes
The changes made in this PR are:
1. Added googlesheets_dataclean.R, under .development
1. Added .development to .Rbuildignore

# Check
- [x] All checks pass
- [x] No errors in googlesheets_dataclean.R
- [x] Code in googlesheets_dataclean.R is well documented
- [x] .development is added to .Rbuildignore

martinctc committed:
feat: Add googlesheets_dataclean.R

Data preparation script for Google Sheets, prior to feeding into Shiny app. Runs once (not run during deployment).

martinctc committed:
chore: Add .development to .Rbuildignore

avisionh commented on .development/googlesheets_dataclean.R:
```diff
@@ -0,0 +1,40 @@
+sheet_url <- "https://docs.google.com/spreadsheets/d/1usk9Q-5lA4bL_z6KXpUohc_2x_KhDgLxtm-YEtim_yk/"
+
+# define google sheets ----------------------------------------------------
+## may consider using purrr once we get to more sheets
```
`purrr` is beautiful

avisionh commented on .development/googlesheets_dataclean.R:
```diff
@@ -0,0 +1,40 @@
+library(tidyverse)
```
Just a note that adding the whole suite of `tidyverse` packages is a lot of overhead as it's not always necessary to use all the packages in `tidyverse`. It's okay to do here as this sits outside the app.

martinctc replied:
Yes, the original intention was that this would just sit outside the app so did not take dependencies etc. into consideration. Since we're adding the packages here to renv, I might replace {tidyverse} with individual packages like {dplyr} and {purrr}. What do you think?

avisionh replied:
Yeah now that you mention it, calling the individual packages will make the installing of packages when running `renv::restore()` faster.

avisionh commented on .development/googlesheets_dataclean.R:
```diff
@@ -0,0 +1,40 @@
+
+clean_gsheet_dat <- function(data){
+  data %>%
+    janitor::clean_names() %>%
```
If you did `clean_names()` instead, would you get the same results?

avisionh replied:
Just curious as to why you need to prepend `clean_names()` since there doesn't seem to be a conflict in multiple functions being called this.

avisionh commented on .development/googlesheets_dataclean.R:
```diff
@@ -0,0 +1,40 @@
+library(tidyverse)
+library(googlesheets4)
```
Could you do `renv::snapshot()` to record the packages in the *renv.lock* file? Also can you add them to the *DESCRIPTION* file as well?

avisionh commented on .development/googlesheets_dataclean.R:
```diff
@@ -0,0 +1,40 @@
+data_master_southern <- googlesheets4::read_sheet(ss = sheet_url, sheet = "Southern")
+
+
+# clean googlesheets ------------------------------------------------------
```
Liking the use of section headers.

avisionh commented on .development/googlesheets_dataclean.R:
```diff
@@ -0,0 +1,40 @@
+sheet_url <- "https://docs.google.com/spreadsheets/d/1usk9Q-5lA4bL_z6KXpUohc_2x_KhDgLxtm-YEtim_yk/"
+
+# define google sheets ----------------------------------------------------
+## may consider using purrr once we get to more sheets
+
+data_master_cnw <- googlesheets4::read_sheet(ss = sheet_url, sheet = "Central & Western")
+data_master_wanchai <- googlesheets4::read_sheet(ss = sheet_url, sheet = "Wan Chai")
+data_master_eastern <- googlesheets4::read_sheet(ss = sheet_url, sheet = "Eastern")
+data_master_southern <- googlesheets4::read_sheet(ss = sheet_url, sheet = "Southern")
```
If you wanted to use `purrr`, it would be like below:
```suggestion
list_arguments <- list(sheet = c("Central & Western", "Wan Chai", "Eastern", "Southern"))
list_data <- pmap(.l = vec_sheet, .f = read_sheet, ss = sheet_url)
names(list_data) <- paste0("data_master_", c("cnw", "wanchai", "eastern", "southern"))
```

avisionh commented on .development/googlesheets_dataclean.R:
```diff
@@ -0,0 +1,40 @@
+# bind list and assign to master ------------------------------------------
+
+master_sheet <-
+  list(data_master_cnw,
+       data_master_wanchai,
+       data_master_eastern,
+       data_master_southern) %>%
```
If you apply `purrr` above in a way akin to what I suggested [above](https://github.com/avisionh/dashboard-hkdistrictcouncillors/pull/8/files#r452510722), then you can shorten this to:
```suggestion
  list_data %>%
```

avisionh reviewed (changes requested):
All passes check but would like `renv::snapshot()` to be run and the packages used in this script added to the *DESCRIPTION* file. 

Rest of my comments are just suggestions if you're interested. :)

martinctc committed:
feat: use purrr for reading in Google Sheets

Update .development/googlesheets_dataclean.R

Co-authored-by: A Ho <a_vision@hotmail.co.uk>

martinctc committed:
feat: use purrr for reading in Google Sheets

Update .development/googlesheets_dataclean.R

Co-authored-by: A Ho <a_vision@hotmail.co.uk>

martinctc committed:
feat: Add DropDownText column

avisionh committed:
fix: Change typo from ZN to EN

This is because it is the English language version of the Chinese name.

avisionh committed:
chore: Remove calls to unnecessary packages

This is to keep the download times and repo smaller.

avisionh committed:
chore: Prompt user to authorise G credentials

This is because we de-authorised in global.R to allow the deployment of app to read the GSheet so need to explicitly authorise again when cleaning the data sheet.

avisionh commented on .development/googlesheets_dataclean.R:
```diff
@@ -0,0 +1,38 @@
+    separate(col = qu_yi_yuan_dc, into = c("DC_ZH", "DC_ZN"), sep = "\n", extra = "merge") %>%
+    rename(ConstituencyCode = "xuan_qu_hao_ma_constituency_code") %>%
+    mutate(DropDownText = paste0(ConstituencyCode, ": ", Constituency_ZH, " / ", Constituency_EN))
+}
```
Should probably hive this off in a folder that contains R scripts for all the functions in the future.

Also, should include unit testing via `testthis` package too.

avisionh committed:
chore: Ignore authorisation keys

This is to avoid exposing to APIs.

avisionh committed:
Merge branch 'master' into feature/data-prep

avisionh reviewed (approved):
Checked the following:
- [x] The code works.
- [x] The packages used are recorded in `renv.lock` and `DESCRIPTION`.
- [x] travis.ci CI/CD and R CMD builds pass.
